File: game/scripts/vscripts/modifiers/upgrades/shards/modifier_upgrade_phantom_dash_damage.lua
modifier_upgrade_phantom_dash_damage = class({})
function modifier_upgrade_phantom_dash_damage:RemoveOnDeath() return false end
function modifier_upgrade_phantom_dash_damage:IsPurgable() return false end

function modifier_upgrade_phantom_dash_damage:OnCreated(params)
    self.range_bonus = 25
    if IsServer() then
        self.damage = 5
        self:SetStackCount(1)
    end
end

function modifier_upgrade_phantom_dash_damage:OnRefresh()
    if IsServer() then
        self:IncrementStackCount()
    end
end

function modifier_upgrade_phantom_dash_damage:GetDamage()
    if IsServer() then
        return self:GetStackCount() * self.damage
    end
end

function modifier_upgrade_phantom_dash_damage:DeclareFunctions()
    return { 
        MODIFIER_PROPERTY_CAST_RANGE_BONUS_STACKING,
    }
end

function modifier_upgrade_phantom_dash_damage:GetModifierCastRangeBonusStacking(params)
    return self:GetStackCount() * self.range_bonus
end

File: src/vscripts/modifiers/upgrades/shards/modifier_upgrade_phantom_dash_damage.ts
import { registerModifier } from "../../../lib/dota_ts_adapter";
import { ModifierUpgrade } from "../modifier_upgrade";

@registerModifier({ customNameForI18n: "modifier_upgrade_phantom_dash_damage" })
export class ModifierUpgradePhantomDashDamage extends ModifierUpgrade {
    rangeBonus = 25;
    damage = 5;

    OnCreated() {
        if (IsServer()) {
            this.SetStackCount(1);
        }
    }

    OnRefresh() {
        if (IsServer()) {
            this.IncrementStackCount();
        }
    }

    GetDamage() {
        if (IsServer()) {
            return this.GetStackCount() * this.damage;
        }
    }

    DeclareFunctions() {
        return [ModifierFunction.CAST_RANGE_BONUS_STACKING];
    }

    GetModifierCastRangeBonusStacking() {
        return this.GetStackCount() * this.rangeBonus;
    }
}
